### PR TITLE
WIP: refactor/hpx

### PR DESCRIPTION
**Pull request by rtohid**

### CODE REVIEW

File: cmake/library.cmake [hunk 1]
```diff
@@ -148,7 +148,7 @@ function(add_target_link_libraries target)
     if ( ${_type} STREQUAL "INTERFACE_LIBRARY")
       target_link_libraries( ${target} INTERFACE ${ARGN} )
     else()
```
Comment (rtohid) [inline cmake/library.cmake]: unnecessary `else`.

File: flecsi/data/CMakeLists.txt [hunk 2]
```diff
@@ -89,37 +90,37 @@ set(data_SOURCES
...
-  SOURCES
-    test/privilege.cc
-)
-
-ftest_add_unit(umesh
-  SOURCES
-    test/umesh.cc
-  THREADS 2
-)
+# ftest_add_unit(canonical
```
Comment (rtohid) [inline flecsi/data/CMakeLists.txt]: Guard with `if(NOT FLECSI_RUNTIME_MODEL STREQUAL "hpx") `.

File: flecsi/execution/CMakeLists.txt [hunk 3]
```diff
@@ -48,6 +48,11 @@ elseif(FLECSI_RUNTIME_MODEL STREQUAL "mpi")
 
 elseif(FLECSI_RUNTIME_MODEL STREQUAL "hpx")
 
+  set(execution_HEADERS
+    ${execution_HEADERS}
```
Comment (rtohid) [inline flecsi/execution/CMakeLists.txt]: Check if it is needed!

File: flecsi/runtime/hpx/context.cc [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+  rp.create_thread_pool("default", hpx::resource::local_priority_fifo);
+
+  // Create a thread pool for executing MPI tasks
+  rp.create_thread_pool("mpi", hpx::resource::static_);
+
+  // Add first core to mpi pool
+  rp.add_resource(rp.numa_domains()[0].cores()[0].pus()[0], "mpi");
+
+  // Now, initialize and run the HPX runtime.
+  return hpx::init();
```
Comment (hkaiser) [inline flecsi/runtime/hpx/context.cc]: We might have to move the initialization to context::initialize, requiring using `hpx::start`/`hpx::stop` instead.

File: flecsi/runtime/hpx/context.hh [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+  }
+
+  /*
+    Documnetation for this interface is in the top-level context type.
+   */
+
+  void set_color(size_t color) {
+    color_ = color;
+  }
+
```
Comment (hkaiser) [inline flecsi/runtime/hpx/context.hh]: Coloring currently assumes that there is one color per index-subtask. We should store that in task-local storage.

File: spack-repo/packages/flecsi/package.py [hunk 6]
```diff
@@ -42,10 +42,11 @@ class Flecsi(CMakePackage):
     depends_on('cmake@3.12:', type='build')
     depends_on('mpi', when='runtime=mpi')
     depends_on('mpi', when='runtime=legion')
-    depends_on('hpx', when='runtime=hpx')
+    depends_on('hpx@master cxxstd=14', when='runtime=hpx')
```
Comment (hkaiser) [inline spack-repo/packages/flecsi/package.py]: make consistent with flecsi master branch

**Comment by hkaiser** [comment]
We should remove the HPX spack package file from the flecsi repo to have it rely on the one in the spack repository.

File: flecsi/execution/hpx/policy.hh [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  class REDUCTION,
+  size_t ATTRIBUTES,
+  typename... ARGS>
+decltype(auto)
+reduce(ARGS &&... args) {
+  using R = typename utils::function_traits<decltype(F)>::return_type;
+
+  hpx::future<R> f = hpx::async(F, std::forward<ARGS>(args)...);
+
+  return f.get();
```
Comment (hkaiser) [inline flecsi/execution/hpx/policy.hh]: This should return the future directly. Will be specialized for `REDUCTION` type.

Comment (hkaiser) [reply flecsi/execution/hpx/policy.hh]: We should enable a (simple) test for this (flecsi/execution/test/task.cc).

File: flecsi/runtime/hpx/context.hh (hunk 5 above)
Comment (hkaiser) [reply flecsi/runtime/hpx/context.hh]: Enable test (flecsi/execution/test/color.cc)

**Comment by hkaiser** [comment]
I think this can be closed now as we started to build the new HPX backend for the devel branch.

**Pull request closed without merging**